packages/big-design/src/components/TableNext/SelectAll/helpers.ts

```diff
@@ -0,0 +1,127 @@
+) {
+  const expandedRowMode = expandedRowSelector !== undefined;
+
+  if (!expandedRowMode) {
+    return [];
+  }
+
+  return expandedRowSelector(parentRow) ?? [];
+}
+
+export function areAllInPageSelected<T>({
+  isExpandable,
+  items,
+  selectedItems,
+  expandedRowSelector,
+}: SelectAllRowsArg<T>) {
+  if (items.length <= 0) {
+    return false;
+  }
+
+  return items.every((parentRow, parentRowIndex) => {
+    const childrenRows: T[] = getChildrenRows(parentRow, expandedRowSelector);
+
+    // Not need to check childrens since expandable mode is not used.
+    if (!isExpandable || childrenRows.length === 0) {
+      return selectedItems[parentRowIndex] !== undefined;
+    }
+
+    return areAllParentsAndChildrenSelected(childrenRows, selectedItems, parentRowIndex);
+  });
+}
+
+export function areSomeInPageSelected<T>({
+  isExpandable,
+  items,
+  selectedItems,
+  expandedRowSelector,
+}: SelectAllRowsArg<T>): boolean {
+  if (items.length <= 0) {
+    return false;
+  }
+
+  return items.some((parentRow, parentRowIndex) => {
+    const childrenRows: T[] = getChildrenRows(parentRow, expandedRowSelector);
+
+    // Not need to check childrens since expandable mode is not used.
+    if (!isExpandable || childrenRows.length === 0) {
+      return selectedItems[parentRowIndex] !== undefined;
+    }
+
+    return areSomeParentsAndChildrenSelected(childrenRows, selectedItems, parentRowIndex);
+  });
+}
+
+function areAllParentsAndChildrenSelected<T>(
+  childrenRows: T[],
+  selectedItems: TableSelectable['selectedItems'],
+  parentRowIndex: number,
+) {
+  const allChildrenRowsSelected = childrenRows.every((_childRow, childRowIndex) => {
+    return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
+  });
+
+  return selectedItems[parentRowIndex] !== undefined && allChildrenRowsSelected;
+}
+
+function areSomeParentsAndChildrenSelected<T>(
+  childrenRows: T[],
+  selectedItems: TableSelectable['selectedItems'],
+  parentRowIndex: number,
+) {
+  const someChildrenRowsInPageSelected = childrenRows.some((_childRow, childRowIndex) => {
+    return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
+  });
+
+  return selectedItems[parentRowIndex] !== undefined && someChildrenRowsInPageSelected;
+}
+
+export function selectAll<T>({
+  expandedRowSelector,
+  isExpandable,
+  items,
+  selectedItems,
+}: SelectAllRowsArg<T>): TableSelectable['selectedItems'] {
+  const newSelectedItems = { ...selectedItems };
+
+  items.forEach((parentRow, parentRowIndex) => {
+    newSelectedItems[`${parentRowIndex}`] = true;
+
+    if (isExpandable) {
+      const childrenRows = getChildrenRows(parentRow, expandedRowSelector);
+
+      childrenRows.forEach((_childRow, childRowIndex) => {
+        if (newSelectedItems[`${parentRowIndex}.${childRowIndex}`] === undefined) {
+          newSelectedItems[`${parentRowIndex}.${childRowIndex}`] = true;
+        }
+      });
+    }
+  });
+
+  return newSelectedItems;
+}
```

Comment:
Do we need to know about previous selected items or can we just build a new object from scratch? This may not be 100% right, but this was what I was thinking:
```suggestion
export function selectAll<T>({
  expandedRowSelector,
  isExpandable,
  items,
}: SelectAllRowsArg<T>): TableSelectable['selectedItems'] {
  return Object.fromEntries(items.map((parent, parentIndex) => {
    const children = getChildrenRows(parent, expandedRowSelector);

    const selectedChildren = isExpandable ? children.map((_child, childIndex) => {
        return [`${parentIndex}.${childIndex}`, true];
    }) : [];

    return [
        [`${parentIndex}`, true],
        ...selectedChildren,
    ];
  }));
}
```